In [61]:
import pandas as pd
import numpy as np
from datetime import datetime


In [62]:
# read csv file
df = pd.read_csv('https://raw.githubusercontent.com/akwasianing/Assignment-2/refs/heads/main/train.csv')
df.head()

Unnamed: 0                              Name    Location  Year  \
0           1  Hyundai Creta 1.6 CRDi SX Option        Pune  2015   
1           2                      Honda Jazz V     Chennai  2011   
2           3                 Maruti Ertiga VDI     Chennai  2012   
3           4   Audi A4 New 2.0 TDI Multitronic  Coimbatore  2013   
4           6            Nissan Micra Diesel XV      Jaipur  2013   

   Kilometers_Driven Fuel_Type Transmission Owner_Type     Mileage   Engine  \
0              41000    Diesel       Manual      First  19.67 kmpl  1582 CC   
1              46000    Petrol       Manual      First    13 km/kg  1199 CC   
2              87000    Diesel       Manual      First  20.77 kmpl  1248 CC   
3              40670    Diesel    Automatic     Second   15.2 kmpl  1968 CC   
4              86999    Diesel       Manual      First  23.08 kmpl  1461 CC   

       Power  Seats  New_Price  Price  
0  126.2 bhp    5.0        NaN  12.50  
1   88.7 bhp    5.0  8.61 Lakh   4.50  
2  88.76 bhp    7.0        NaN   6.00  
3  140.8 bhp    5.0        NaN  17.74  
4   63.1 bhp    5.0        NaN   3.50

In [63]:
#  drop the unnamed column as it is not a relevant feature
df_cleaned = df.drop(columns=['Unnamed: 0'])
df_cleaned.head()


Name    Location  Year  Kilometers_Driven  \
0  Hyundai Creta 1.6 CRDi SX Option        Pune  2015              41000   
1                      Honda Jazz V     Chennai  2011              46000   
2                 Maruti Ertiga VDI     Chennai  2012              87000   
3   Audi A4 New 2.0 TDI Multitronic  Coimbatore  2013              40670   
4            Nissan Micra Diesel XV      Jaipur  2013              86999   

  Fuel_Type Transmission Owner_Type     Mileage   Engine      Power  Seats  \
0    Diesel       Manual      First  19.67 kmpl  1582 CC  126.2 bhp    5.0   
1    Petrol       Manual      First    13 km/kg  1199 CC   88.7 bhp    5.0   
2    Diesel       Manual      First  20.77 kmpl  1248 CC  88.76 bhp    7.0   
3    Diesel    Automatic     Second   15.2 kmpl  1968 CC  140.8 bhp    5.0   
4    Diesel       Manual      First  23.08 kmpl  1461 CC   63.1 bhp    5.0   

   New_Price  Price  
0        NaN  12.50  
1  8.61 Lakh   4.50  
2        NaN   6.00  
3        NaN  17.74  
4        NaN   3.50

In [64]:
df_cleaned.shape

(5847, 13)

In [65]:
df_cleaned.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 5847 entries, 0 to 5846
Data columns (total 13 columns):
 #   Column             Non-Null Count  Dtype  
---  ------             --------------  -----  
 0   Name               5847 non-null   object 
 1   Location           5847 non-null   object 
 2   Year               5847 non-null   int64  
 3   Kilometers_Driven  5847 non-null   int64  
 4   Fuel_Type          5847 non-null   object 
 5   Transmission       5847 non-null   object 
 6   Owner_Type         5847 non-null   object 
 7   Mileage            5845 non-null   object 
 8   Engine             5811 non-null   object 
 9   Power              5811 non-null   object 
 10  Seats              5809 non-null   float64
 11  New_Price          815 non-null    object 
 12  Price              5847 non-null   float64
dtypes: float64(2), int64(2), object(9)
memory usage: 594.0+ KB


In [66]:
# extract numeric values from attributes (Mileage, Engine, Power, New_Price) and remove units

df_cleaned['Mileage'] = df_cleaned['Mileage'].str.extract(r'(\d+\.?\d*)').astype(float)

df_cleaned['Engine'] = df_cleaned['Engine'].str.extract(r'(\d+\.?\d*)').astype(float)

df_cleaned['Power'] = df_cleaned['Power'].str.extract(r'(\d+\.?\d*)').astype(float)

df_cleaned['New_Price'] = df_cleaned['New_Price'].str.extract(r'(\d+\.?\d*)').astype(float)



In [67]:
# Add units to column names for clarity

column_units = {
    'Mileage': 'Mileage (kmpl)',
    'Engine': 'Engine (cc)',
    'Power': 'Power (bhp)',
    'New_Price': 'New_Price (Lakh)'
}

# Rename the columns
df_cleaned.rename(columns=column_units, inplace=True)

In [68]:
df_cleaned.head(10)

Name    Location  Year  Kilometers_Driven  \
0     Hyundai Creta 1.6 CRDi SX Option        Pune  2015              41000   
1                         Honda Jazz V     Chennai  2011              46000   
2                    Maruti Ertiga VDI     Chennai  2012              87000   
3      Audi A4 New 2.0 TDI Multitronic  Coimbatore  2013              40670   
4               Nissan Micra Diesel XV      Jaipur  2013              86999   
5    Toyota Innova Crysta 2.8 GX AT 8S      Mumbai  2016              36000   
6  Volkswagen Vento Diesel Comfortline        Pune  2013              64430   
7       Tata Indica Vista Quadrajet LS     Chennai  2012              65932   
8                     Maruti Ciaz Zeta       Kochi  2018              25692   
9          Honda City 1.5 V AT Sunroof     Kolkata  2012              60000   

  Fuel_Type Transmission Owner_Type  Mileage (kmpl)  Engine (cc)  Power (bhp)  \
0    Diesel       Manual      First           19.67       1582.0       126.20   
1    Petrol       Manual      First           13.00       1199.0        88.70   
2    Diesel       Manual      First           20.77       1248.0        88.76   
3    Diesel    Automatic     Second           15.20       1968.0       140.80   
4    Diesel       Manual      First           23.08       1461.0        63.10   
5    Diesel    Automatic      First           11.36       2755.0       171.50   
6    Diesel       Manual      First           20.54       1598.0       103.60   
7    Diesel       Manual     Second           22.30       1248.0        74.00   
8    Petrol       Manual      First           21.56       1462.0       103.25   
9    Petrol    Automatic      First           16.80       1497.0       116.30   

   Seats  New_Price (Lakh)  Price  
0    5.0               NaN  12.50  
1    5.0              8.61   4.50  
2    7.0               NaN   6.00  
3    5.0               NaN  17.74  
4    5.0               NaN   3.50  
5    8.0             21.00  17.50  
6    5.0               NaN   5.20  
7    5.0               NaN   1.95  
8    5.0             10.65   9.95  
9    5.0               NaN   4.49

In [69]:
# show info after removing units from values

df_cleaned.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 5847 entries, 0 to 5846
Data columns (total 13 columns):
 #   Column             Non-Null Count  Dtype  
---  ------             --------------  -----  
 0   Name               5847 non-null   object 
 1   Location           5847 non-null   object 
 2   Year               5847 non-null   int64  
 3   Kilometers_Driven  5847 non-null   int64  
 4   Fuel_Type          5847 non-null   object 
 5   Transmission       5847 non-null   object 
 6   Owner_Type         5847 non-null   object 
 7   Mileage (kmpl)     5845 non-null   float64
 8   Engine (cc)        5811 non-null   float64
 9   Power (bhp)        5811 non-null   float64
 10  Seats              5809 non-null   float64
 11  New_Price (Lakh)   815 non-null    float64
 12  Price              5847 non-null   float64
dtypes: float64(6), int64(2), object(5)
memory usage: 594.0+ KB


In [70]:
# identify missing values in all columns

df_cleaned.isna().sum()

Name                    0
Location                0
Year                    0
Kilometers_Driven       0
Fuel_Type               0
Transmission            0
Owner_Type              0
Mileage (kmpl)          2
Engine (cc)            36
Power (bhp)            36
Seats                  38
New_Price (Lakh)     5032
Price                   0
dtype: int64

In [71]:
# Mileage column - drop rows containing missing values

df_cleaned = df_cleaned.dropna(subset=['Mileage (kmpl)'])

For mileage column since only 2 out of 5847 rows have missing values, these two rows can be dropped. (negligible impact)

In [72]:
# Imputing missing values in the columns (Engine, Power, Seats)
# Calculate skewness to decide imputation method

df_cleaned[['Engine (cc)', 'Power (bhp)', 'Seats']].skew()

Engine (cc)    1.412804
Power (bhp)    1.915841
Seats          1.911640
dtype: float64

Since skewness values for the columns (Engine, Power, Seats) are positive, indicating right-skewed distributions, the missing values in these columns are replaced with their respective medians, as the median is robust to outliers and better suited for skewed data



In [73]:
# Calculate the median of each specified column

medians = df_cleaned[['Engine (cc)', 'Power (bhp)', 'Seats']].median()
print(medians)

Engine (cc)    1497.0
Power (bhp)      98.6
Seats             5.0
dtype: float64


In [74]:
# fill in the missing values with the calculated medians

df_cleaned.loc[:, ['Engine (cc)', 'Power (bhp)', 'Seats']] = df_cleaned.loc[:, ['Engine (cc)', 'Power (bhp)', 'Seats']].fillna(medians)


In [75]:
# drop New_Price column

df_cleaned = df_cleaned.drop(['New_Price (Lakh)'], axis=1)

New_Price column is dropped due to a high percentage of missing values (~86%). With such a significant portion of data missing, retaining the column would likely reduce the quality and interpretability of the analysis. Imputation would rather be unreliable and potentially introduce bias that is why the New_Price column is dropped.


In [76]:
# Verify if missing values for the columns (Engine, Power, Seats) are filled with respective medians; the missing values in the Mileage is dropped and the entire New_Price column is dropped.
df_cleaned.isna().sum()

Name                 0
Location             0
Year                 0
Kilometers_Driven    0
Fuel_Type            0
Transmission         0
Owner_Type           0
Mileage (kmpl)       0
Engine (cc)          0
Power (bhp)          0
Seats                0
Price                0
dtype: int64

In [77]:
# make a copy of df_cleaned data

df_encoded = df_cleaned.copy()


In [78]:
# convert 'Fuel_Type' and 'Transmission' into numerical one-hot encoded columns
# assuming using models not sensitive to multicollinearity such as decision trees

df_encoded = pd.get_dummies(df_cleaned, columns=['Fuel_Type', 'Transmission'], dtype=int)


In [79]:
df_encoded.head()


Name    Location  Year  Kilometers_Driven  \
0  Hyundai Creta 1.6 CRDi SX Option        Pune  2015              41000   
1                      Honda Jazz V     Chennai  2011              46000   
2                 Maruti Ertiga VDI     Chennai  2012              87000   
3   Audi A4 New 2.0 TDI Multitronic  Coimbatore  2013              40670   
4            Nissan Micra Diesel XV      Jaipur  2013              86999   

  Owner_Type  Mileage (kmpl)  Engine (cc)  Power (bhp)  Seats  Price  \
0      First           19.67       1582.0       126.20    5.0  12.50   
1      First           13.00       1199.0        88.70    5.0   4.50   
2      First           20.77       1248.0        88.76    7.0   6.00   
3     Second           15.20       1968.0       140.80    5.0  17.74   
4      First           23.08       1461.0        63.10    5.0   3.50   

   Fuel_Type_Diesel  Fuel_Type_Petrol  Transmission_Automatic  \
0                 1                 0                       0   
1                 0                 1                       0   
2                 1                 0                       0   
3                 1                 0                       1   
4                 1                 0                       0   

   Transmission_Manual  
0                    1  
1                    1  
2                    1  
3                    0  
4                    1

In [80]:
# creating one more attibute (Car_Age) and adding to dataset

current_year = datetime.now().year
df_encoded['Car_Age'] = current_year - df_encoded['Year']
df_encoded.head()


Name    Location  Year  Kilometers_Driven  \
0  Hyundai Creta 1.6 CRDi SX Option        Pune  2015              41000   
1                      Honda Jazz V     Chennai  2011              46000   
2                 Maruti Ertiga VDI     Chennai  2012              87000   
3   Audi A4 New 2.0 TDI Multitronic  Coimbatore  2013              40670   
4            Nissan Micra Diesel XV      Jaipur  2013              86999   

  Owner_Type  Mileage (kmpl)  Engine (cc)  Power (bhp)  Seats  Price  \
0      First           19.67       1582.0       126.20    5.0  12.50   
1      First           13.00       1199.0        88.70    5.0   4.50   
2      First           20.77       1248.0        88.76    7.0   6.00   
3     Second           15.20       1968.0       140.80    5.0  17.74   
4      First           23.08       1461.0        63.10    5.0   3.50   

   Fuel_Type_Diesel  Fuel_Type_Petrol  Transmission_Automatic  \
0                 1                 0                       0   
1                 0                 1                       0   
2                 1                 0                       0   
3                 1                 0                       1   
4                 1                 0                       0   

   Transmission_Manual  Car_Age  
0                    1       10  
1                    1       14  
2                    1       13  
3                    0       12  
4                    1       12

In [81]:
#  select specified columns

df_encoded_selected = df_encoded[['Name', 'Year', 'Price']]
df_encoded_selected.head()

Name  Year  Price
0  Hyundai Creta 1.6 CRDi SX Option  2015  12.50
1                      Honda Jazz V  2011   4.50
2                 Maruti Ertiga VDI  2012   6.00
3   Audi A4 New 2.0 TDI Multitronic  2013  17.74
4            Nissan Micra Diesel XV  2013   3.50

In [82]:
# filter rows where the car price is greater than 35 Lakh

df_encoded_filtered = df_encoded[df_encoded['Price'] > 35]
df_encoded_filtered.head()

Name    Location  Year  \
65   Mercedes-Benz C-Class Progressive C 220d  Coimbatore  2019   
87           Audi A6 2011-2015 35 TDI Premium  Coimbatore  2018   
127                  Mercedes-Benz SLC 43 AMG     Kolkata  2017   
140                            Audi RS5 Coupe      Mumbai  2013   
160                         BMW X5 xDrive 30d      Mumbai  2015   

     Kilometers_Driven Owner_Type  Mileage (kmpl)  Engine (cc)  Power (bhp)  \
65               15369      First            0.00       1950.0       194.00   
87               48367      First           17.68       1968.0       174.33   
127              13372      First           19.00       2996.0       362.07   
140              23000      First           11.05       2894.0       444.00   
160              41000      First           15.30       2993.0       258.00   

     Seats  Price  Fuel_Type_Diesel  Fuel_Type_Petrol  Transmission_Automatic  \
65     5.0  35.67                 1                 0                       1   
87     5.0  39.58                 1                 0                       1   
127    2.0  54.00                 0                 1                       1   
140    4.0  37.00                 0                 1                       1   
160    5.0  45.00                 1                 0                       1   

     Transmission_Manual  Car_Age  
65                     0        6  
87                     0        7  
127                    0        8  
140                    0       12  
160                    0       10

In [83]:
# rename columns: 'Name' → 'Make_and_Model', 'Price' → 'Original_Price'

df_encoded_renamed = df_encoded.rename(columns={'Name': 'Make_and Model', 'Price': 'Original_Price'})
df_encoded_renamed.head()


Make_and Model    Location  Year  Kilometers_Driven  \
0  Hyundai Creta 1.6 CRDi SX Option        Pune  2015              41000   
1                      Honda Jazz V     Chennai  2011              46000   
2                 Maruti Ertiga VDI     Chennai  2012              87000   
3   Audi A4 New 2.0 TDI Multitronic  Coimbatore  2013              40670   
4            Nissan Micra Diesel XV      Jaipur  2013              86999   

  Owner_Type  Mileage (kmpl)  Engine (cc)  Power (bhp)  Seats  Original_Price  \
0      First           19.67       1582.0       126.20    5.0           12.50   
1      First           13.00       1199.0        88.70    5.0            4.50   
2      First           20.77       1248.0        88.76    7.0            6.00   
3     Second           15.20       1968.0       140.80    5.0           17.74   
4      First           23.08       1461.0        63.10    5.0            3.50   

   Fuel_Type_Diesel  Fuel_Type_Petrol  Transmission_Automatic  \
0                 1                 0                       0   
1                 0                 1                       0   
2                 1                 0                       0   
3                 1                 0                       1   
4                 1                 0                       0   

   Transmission_Manual  Car_Age  
0                    1       10  
1                    1       14  
2                    1       13  
3                    0       12  
4                    1       12

In [84]:
# mutate - create new column for original car price in USD assuming 1 Lakh is 1200 USD

df_encoded_renamed['Original_Price_USD'] = df_encoded_renamed['Original_Price'] * 1200
df_encoded_renamed.head()

Make_and Model    Location  Year  Kilometers_Driven  \
0  Hyundai Creta 1.6 CRDi SX Option        Pune  2015              41000   
1                      Honda Jazz V     Chennai  2011              46000   
2                 Maruti Ertiga VDI     Chennai  2012              87000   
3   Audi A4 New 2.0 TDI Multitronic  Coimbatore  2013              40670   
4            Nissan Micra Diesel XV      Jaipur  2013              86999   

  Owner_Type  Mileage (kmpl)  Engine (cc)  Power (bhp)  Seats  Original_Price  \
0      First           19.67       1582.0       126.20    5.0           12.50   
1      First           13.00       1199.0        88.70    5.0            4.50   
2      First           20.77       1248.0        88.76    7.0            6.00   
3     Second           15.20       1968.0       140.80    5.0           17.74   
4      First           23.08       1461.0        63.10    5.0            3.50   

   Fuel_Type_Diesel  Fuel_Type_Petrol  Transmission_Automatic  \
0                 1                 0                       0   
1                 0                 1                       0   
2                 1                 0                       0   
3                 1                 0                       1   
4                 1                 0                       0   

   Transmission_Manual  Car_Age  Original_Price_USD  
0                    1       10             15000.0  
1                    1       14              5400.0  
2                    1       13              7200.0  
3                    0       12             21288.0  
4                    1       12              4200.0

In [85]:
# arrange - sort Original_Price in USD in descending order

df_encoded_sort = df_encoded_renamed.sort_values(by='Original_Price_USD', ascending=False)
df_encoded_sort.head()

Make_and Model    Location  Year  \
3952  Land Rover Range Rover 3.0 Diesel LWB Vogue   Hyderabad  2017   
5620                   Lamborghini Gallardo Coupe       Delhi  2011   
5752                       Jaguar F Type 5.0 V8 S   Hyderabad  2015   
1457              Land Rover Range Rover Sport SE       Kochi  2019   
1917                           BMW 7 Series 740Li  Coimbatore  2018   

      Kilometers_Driven Owner_Type  Mileage (kmpl)  Engine (cc)  Power (bhp)  \
3952              25000      First           13.33       2993.0        255.0   
5620               6500      Third            6.40       5204.0        560.0   
5752               8000      First           12.50       5000.0        488.1   
1457              26013      First           12.65       2993.0        255.0   
1917              28060      First           12.05       2979.0        320.0   

      Seats  Original_Price  Fuel_Type_Diesel  Fuel_Type_Petrol  \
3952    5.0          160.00                 1                 0   
5620    2.0          120.00                 0                 1   
5752    2.0          100.00                 0                 1   
1457    5.0           97.07                 1                 0   
1917    5.0           93.67                 0                 1   

      Transmission_Automatic  Transmission_Manual  Car_Age  Original_Price_USD  
3952                       1                    0        8            192000.0  
5620                       1                    0       14            144000.0  
5752                       1                    0       10            120000.0  
1457                       1                    0        6            116484.0  
1917                       1                    0        7            112404.0

In [86]:
# summarise with group by operation - Average Price by Location (City)

df_encoded_renamed.groupby('Location')['Original_Price'].mean().reset_index(name='Avg_Price')

Location  Avg_Price
0    Ahmedabad   8.567248
1    Bangalore  13.482670
2      Chennai   7.947726
3   Coimbatore  15.160206
4        Delhi   9.881944
5    Hyderabad   9.997423
6       Jaipur   5.916725
7        Kochi  11.309109
8      Kolkata   5.733924
9       Mumbai   9.588397
10        Pune   6.951000

In [87]:
# summarise with group by operation - count of Cars by City


df_encoded_renamed.groupby(['Location']).size().reset_index(name='Car_Count')

Location  Car_Count
0    Ahmedabad        218
1    Bangalore        352
2      Chennai        475
3   Coimbatore        631
4        Delhi        540
5    Hyderabad        710
6       Jaipur        403
7        Kochi        640
8      Kolkata        525
9       Mumbai        761
10        Pune        590